[grammar] 'override' keyword

sailingKieler opened the pull request:
Updated!

Introduces `@Override` as desired by #121.
* added a corresponding attribute to `AbstractRule` in `Xtext.ecore`
* added an optional leading keyword '@Override' in the rules `ParserRule`, `TerminalRule`, and `EnumRule`

Comments, suggestions, and approvals are welcome :-)

meysholdt commented on org.eclipse.xtext/src/org/eclipse/xtext/Xtext.xtext:
```diff
@@ -251,6 +253,7 @@ CharacterRange returns AbstractElement:
 ;
 
 EnumRule:
+	(^override?='@Override')?
```
`^` is now not necessary anymore

sailingKieler replied:
probably not

svenefftinge commented:
Sorry, if my proposal wasn't clear. I meant to introduce the notion of annotations in the grammar.
So a rule would be preceded by any number of annotations.
Something like this:

```
ParserRule :
   (annotations+=Annotation)*
 ....

Annotation : 
   '@' name=ID
;
```

svenefftinge commented:
Would it be a lot of work to do the bootstrap stuff separately?

sailingKieler commented:
No I don't think so. I can cherry-pick the grammar change so the bootstrapping stuff can be merged separately.

sailingKieler commented:
Regarding the explicit annotations:
I got it and agree, but we doubt we can achieve that solution in the currently available time, so we prefer the interim solution.

svenefftinge commented on org.eclipse.xtext/src/org/eclipse/xtext/xtext/XtextValidator.java:
```diff
@@ -1234,5 +1234,32 @@ public void checkOppositeReferenceUsed(Assignment assignment) {
+				
+			} else {
+				if (rule.isOverride()) {
+					error("None of the used grammars contain a rule named " + name, rule,
```
"used grammars" is not a user-level term. 
How about "The rule 'Foo' must override a rule from a super grammar"?

svenefftinge commented:
Maybe I'm missing something, but it shouldn't be more work.
You'd basically need to implement `isDeprecated` which internally checks for name=='Override' but not much more.

svenefftinge commented:
Actually, I would like to see the "SuppressWarnings[code1, code2]" thing Sebastian added one year ago to be replaced with proper annotations, too.
See https://github.com/eclipse/xtext-core/blob/master/org.eclipse.xtext/src/org/eclipse/xtext/xtext/XtextDiagnosticConverter.java#L94

knutwannheden commented:
Making `Xtext.xtext` inherit from `XbaseWithAnnotations.xtext` might propel the bootstrapping to the next level of fun! ;-)

sailingKieler commented:
@knutwannheden for sure! :-)

sailingKieler commented:
Just pushed my updated draft.
`AbstractRule` now has a list of `Annotation`s, we however permit just one annotation per rule in the concrete syntax for now. A dedicated check ensures that only `@Override` is accepted.

sailingKieler commented:
Removed the bootstrap project from the changes of this PR.
Added corresponding test in `XtextValidatorTest` testing the four scenarios:
* no rule overriding + no annotation
* rule overriding + no annotation
* no rule overriding + annotation
* rule overriding + annotation

sailingKieler commented:
From my point of view, the PR can be merged,
and afterwards the branches 'cs/override' on xtext-eclipse + xtext-idea as well.

sailingKieler commented on org.eclipse.xtext.tests/src/org/eclipse/xtext/xtext/XtextValidationTest.java:
```diff
@@ -102,6 +102,42 @@ private void configureValidator(XtextValidator validator, ValidationMessageAccep
+		List<Issue> issues3 = validator.validate(resource3, CheckMode.FAST_ONLY, CancelIndicator.NullImpl);
+		assertEquals(issues3.toString(), 0, issues3.size());
+	}
+	
```
Sure.

sailingKieler replied:
Done.

meysholdt commented on org.eclipse.xtext/src/org/eclipse/xtext/xtext/XtextValidator.java:
```diff
@@ -1234,5 +1235,45 @@ public void checkOppositeReferenceUsed(Assignment assignment) {
+			
+			if (r != null) {
+				if (!isOverride) {
+					warning("This rule overrides " + name + " in " + GrammarUtil.getGrammar(r).getName() + ".", rule,
```
* The warning message should also indicate what to do about it:
`This rule overrides Foo in Bar and thus should be annotated with @Override".

* please add an Issue Code, so that later a QuickFix can be provided.

svenefftinge replied:
+1

meysholdt commented:
@svenefftinge could you review this PR?

svenefftinge commented on org.eclipse.xtext/src/org/eclipse/xtext/xtext/XtextValidator.java:
```diff
@@ -1234,5 +1235,45 @@ public void checkOppositeReferenceUsed(Assignment assignment) {
-
+	
+	@Check
+	public void checkAnnotations(AbstractRule rule) {
```
We should remove this check, because the reason I asked for generic annotations is to allow users to add arbitrary ones without changing our code. E.g. to use them in custom generator fragments.

knutwannheden replied:
How about an error validation that can be disabled in the preferences (or even using a system property)? Because some users would certainly be willing to add the declarations once that is required. Then they could at least already benefit from the new syntax. For many use cases (like @SuppressWarnings) that would of course also require some syntax for parameters.

meysholdt replied:
@svenefftinge and I discussed offline that it would be OK for now to allow arbitrary annotations without requiring annotations declarations. In the future, if we choose to introduce annotations declarations, we can offer a smooth migration path by rising a warning (not an error!) for annotations without declarations.

sailingKieler commented:
Updated the PR wrt. the requested changes.

sailingKieler commented:
Just fixed the test I forgot to update while fixing the error msg.
